flink-runtime/src/main/java/org/apache/flink/runtime/taskexecutor/TaskExecutor.java

```diff
@@ -851,8 +852,8 @@ public void heartbeatFromResourceManager(ResourceID resourceID) {
 	}
 
 	@Override
-	public CompletableFuture<SerializableOptional<String>> requestMetricQueryServiceAddress(Time timeout) {
-		return CompletableFuture.completedFuture(SerializableOptional.ofNullable(metricQueryServicePath));
+	public CompletableFuture<Optional<MetricQueryServiceGateway>> requestMetricQueryServiceGateway(Time timeout) {
```

Comment:
This doesn't work at runtime since `Optional` is not serializable. Investigating alternatives.

Reply:
Is using `SerializableOptional` reasonable? It causes we implements `Serializable` for `MetricQueryServiceGateway` and thus for `MetricQueryService`.

Reply:
You can't just add `Serializable` to a class and expect it to magically work. The `MQS` extends `RpcEndpoint` which is not serialiable and contains loads of fields that are neither.

From what i can tell we have to return the RpcService address instead, but again, let me investigate first before making any changes.

Reply:
Then the behavior should be like we previously do with `actorSelection`. Thanks for your help and I would also learn how to pass an address and connect to the RpcEndpoint with it.

Reply:
yes it is very similar to the previous behavior.